x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java

```diff
@@ -165,12 +173,46 @@ private Object unwrapMultiValue(Object values) {
                 return DateUtils.asDateTime(Long.parseLong(values.toString()), zoneId);
             }
         }
+        
         // The Jackson json parser can generate for numerics - Integers, Longs, BigIntegers (if Long is not enough)
         // and BigDecimal (if Double is not enough)
-        if (values instanceof Number
-                || values instanceof String
-                || values instanceof Boolean) {
-            return values;
+        if (values instanceof Number || values instanceof String || values instanceof Boolean) {
+            if (dataType == null) {
+                return values;
+            }
+            if (dataType.isNumeric() && dataType.isFromDocValuesOnly() == false) {
+                if (dataType == DataType.DOUBLE || dataType == DataType.FLOAT || dataType == DataType.HALF_FLOAT) {
+                    return NumberType.DOUBLE.parse(values, true);
+                } else {
+                    long l = (long) NumberType.LONG.parse(values, true);
+                    switch (dataType) {
+                        case BYTE:
+                            if (l > Byte.MAX_VALUE || l < Byte.MIN_VALUE) {
+                                return l;
```

Comment:
Could you please double check if these ifs are needed (in terms of the error message produced, since the parse() itself does internally this validation)?
If they're still valid, please add a comment to indicate this.